% !TEX root = snp-contam-main.Rnw

\chapter{Introduction}

Single nucleotide polymorphisms (SNPs), which are locations in the DNA sequence that vary by a single base pair, have emerged as an important genetic marker, and their use is becoming widespread in molecular ecology.  The low cost of SNP genotyping makes it a practical alternative to microsatellite genotyping \citep{Anderson&Garza2006}.  SNP genotyping codes are also independent of the genotyping system, allowing for easy standardization across laboratories \citep{Morinetal2004}.  In addition, SNP loci (or gene locations in the genome) are present in large amounts in many species' genotypes \citep{Morinetal2004}, specific SNP loci under selection can be divergent between populations \citep{Freamoetal2011, Karlssonetal2011}, and SNP genotyping can be used on short fragments, which allows for low genotyping failure and error rates in historical or low-quality samples \citep{Morin&McCarthy2007, Smithetal2011}.  Studies have also shown that SNPs are comparable to or better than microsatellites in most applications.  According to several studies, both microsatellites and SNPs are effective markers for population assignment and genetic stock identification, although a large number of SNPs may be required to identify fine-scale population differentiation \citep{Narumetal2008, Hessetal2011, Morinetal2004}.  SNPs have also been shown outperform microsatellites in the accuracy of parentage assignments \citep{Hauseretal2011}, and the effectiveness and accuracy of linkage maps for chromosomes \citep{Balletal2010}.

However, one disadavantage to SNP genotyping is its lack of a standardized processes for identifying contamination.  Cross-contaminated DNA samples, which contain DNA from two or more individuals, can confound results of population assignment, genetic stock identification, parentage assignment, and other genetic application.  It is therefore important that contamination is removed before analysis. Identifying contaminated samples using SNP genotyping data is more difficult than with genotyping data from microsatellites.  Because microsatellites are highly polymorphic, meaning that each microsatellite loci has many different alleles, contamination is detected when there are more than two alleles in a diploid sample.  This method is not possible with SNPs given their bi-allelic nature; even a contaminated sample will have only one or two different alleles.  A logical method to eliminate contamination is to determine an individual's percentage of heterozygous loci, which is the percentage of loci that have two different alleles.  Theoretically, contaminated samples will have an excess of heterozygous loci, as the presence of DNA from two or more individuals will increase the rate of heterozygousity.  Although considering heterozygousity does not use all the information in the SNP genotyping data, it can be a powerful method for simple tests of single individuals.  However, in the presence of multiple tests, a principled modeling of the situation is preferable.

Several studies have developed models using SNP genotyping data to identify contamination in human DNA samples \citep{Junetal2012, Cibulskisetal2011, Huangetal2014}.  However, these models incorporate sequencing data as well as information from the complete human genome, which is not available for non-model organisms studied in molecular ecology.  Jasper et. al (2012) developed a Bayesian model to identify cross-contaminated DNA samples in salmon.  This model, however, required an estimate of the rate of contamination to perform selection between contaminated and uncontaminated genotype score probabilities; in many cases, it may not be possible to provide a reasonable estimate for the contamination rate.

This paper presents a simple Bayesian model that uses Markov Chain Monte Carlo (MCMC) methods to identify contaminated samples given an unknown rate of contamination.   The model will be most useful in situations where laboratory error has accidentally pooled two DNA samples.  The model is tailored for two applications of SNP genotyping: (1) the estimation of allele frequencies of a single population and (2) the population identification of organisms sampled from a mixture of multiple populations for which reference samples are available.  This paper describes the models and MCMC algorithms used to sample from the posterior distribution of our unknown variables as well as  a set of simulations used to assess the model and demonstrate the model's accurate identification of contaminated samples.

\chapter{Background}
\section{Baysian Inference}
Bayesian inference combines new evidence with prior knowledge to determine the sequence of events that most likely led to a given outcome.  Equation~\ref{bayes} gives the formula for Bayes' rule.  

\begin{equation} \label{bayes}
p(\theta|y) = \frac{p(y|\theta)p(\theta)}{p(y)}
\end{equation}

In Bayesian inference, Bayes' rule is used to update prior beliefs and acquire a new \emph{posterior distribution}, $p(\theta|y)$.  The variable $\theta$ represents an unknown, like the allele frequencies at a set of loci, and the variable $y$ is the known paramenter, like an individuals genotype.

The prior knowledge is expressed in Bayes'rule by $p(y)$, which is known as the \emph{prior distribution}.  New information gained from the known parameter $y$ is incorporated into Bayes' rule with $p(y|\theta)$. $p(y|\theta)$, which is known as the \emph{likelihood}, is a function of $\theta$.  The denominator of Bayes' rule is given by $p(y) = \int p(y|\theta) p(\theta) d\theta$.  Equation~\ref{bayes} is not always easily solved, but algorithms, like MCMC methods, can be used to perform inference using Bayes'rule.

\section{Markov Chain Monte Carlo (MCMC)}
Markov chain Monte Carlo methods are algorithms that sample from probability distribution that follows the stochastic prosses of a Markov chain.  The equilibrium of this probability distribution is the desired true posterior distribution, and MCMC methods can therefore be used to estimate values from the true distribution.
First, a Monte Carlo method approximates an expectation using the mean of a sample of random variables.  It relies on the weak law of large numbers, which is given by Equation~\ref{wlln} where $X_1, X_2, \ldots, X_n$ are independent and identically distributed random variables with $\mathbb{E}[X_i] < \infty$ and $\bar{X_n}=\frac{1}{n}\sum_{i=1}^{n}X_1$.
\begin{equation} \label{wlln}
\lim_{n \to \infty} P(|\bar{X_n}-\mathbb{E}[X_i]|>\epsilon) = 0
\end{equation}

The weak law of large numbers leads to the conclusion that an expectation can be approximated by the sample mean of $n$ random variables (Equation~\ref{expectation}).
\begin{equation} \label{expectation}
\mathbb{E}[g(X)] \approx \frac{1}{n} \sum_{i=1}^{n} g(x^{(i)})
\end{equation}

In an MCMC algorithm, $x^{(i)}$ are the states visited by a Markov chain.  Markov chains are defined by Equation~\ref{markov} and depicted schematically by Figure~\ref{markovchain} where $X_t$ is a sequence of random variables for $t=0,1,2,\ldots$.
\begin{equation} \label{markov}
P(X_i|X_{t-1},X_{t-2},\ldots, X_0) = P(X_t|X_{t-1})  \forall t
\end{equation}

\begin{figure}[h]
\begin{center}
\includegraphics[width=2.5in]{./images/Markov_chain.pdf}
\end{center}
\caption{Schematic diagram of a Markov chain.}
\label{markovchain}
\end{figure}

In order for an MCMC to approximate the correct values, a Markov chain must have the correct limiting distribution and it must be \emph{ergodic}.  A Markov chain is \emph{ergodic} if it has a \emph{limiting distribution} $\pi$.  According to the Ergodic Theorem, a Markov chain will be ergodic if it has no transient states, if it is aperiodic, and if it is irreducible.  The MCMC algorithms in this paper use Gibbs Sampling to ensure that these properties are satisfied and that the random sequence is truly a Markov chain.

\section{Gibbs Sampling}
The Gibbs sampler generates samples from a joint probability distribution by using the \emph{full conditional} distribution for each parameter.  If $p(\theta_1, \theta_2, \theta_3, \ldots, \theta_n)$ is the joint distribution, for example, the full conditional distrubtion of $\theta_1$ would be only the conditional distribution of $\theta_1$ given all other variables and the known information or $p(\theta_1|\theta_{-1},y)$, where $y$ is the known information. Values for each parameter, $(\theta_1, \theta_2, \theta_3, \ldots, \theta_n)$, can be updated by drawing from their respective full conditional distributions.  After $N$ number of repetitions, the drawn values will be Markov chains that can be used to approximate the expectations of each distribution.   

%Section on the SSP Model
\chapter{Model Derivation}
The MCMC models for identifying contaminated samples were developed and executed in R (v3.1.0).  The sections below describe the development of the MCMC algorithms used in the R code.

\section{Single Source Population Model}
This section describes the formulation of the Single Source Population (SSP) model, which is used for detecting contaminated samples from a collection in which all the
individuals, as well as any contaminating DNA, originate from a single population.  The collection of DNA in the SSP model
includes $N$ genetic samples, and each sample is genotyped at $L$ SNP loci, which are assumed to not be in linkage 
disequilibrium (LD) in the population.  One allele at each SNP is labeled ``0'' 
and the other ``1''.  At the $\ell\thh$ SNP, $\theta_\ell$ denotes the frequency of the ``1'' allele, 
$\ell\in\{1,\ldots,L\}$, in the population from which the samples originate. Associated with each sample
$i$ ($i\in\{1,\ldots,N\})$ is an unobserved indicator variable $z_i$ which takes the value 1 if sample $i$ 
is contaminated (contains DNA from 2 individuals) and 0 if sample $i$ contains DNA from just a single 
individual. The known variable $y_{i,\ell}$ denotes the observed genotype of sample $i$ at locus $\ell$.  If a sample 
contains only ``0'' alleles, $y_{i,\ell}=0$, if both a ``0'' and ``1'' allele are detected, $y_{i,\ell}
=1$, and if the sample contains only ``1'' alleles then $y_{i,\ell}=2$.  If sample $i$ is not 
contaminated, and is from a diploid, the probability of $y_{i,\ell}$ is assumed to follow independent 
sampling of alleles from the population, conforming to the Hardy-
Weinberg law:
\begin{eqnarray} \label{HW_clean}
P(y_{i,\ell} = 0\:|\:z_i = 0) & = & (1-\theta_\ell)^2 \nonumber\\
P(y_{i,\ell} = 1\:|\:z_i = 0) & = & 2\theta_\ell(1-\theta_\ell) \\
P(y_{i,\ell} = 2\:|\:z_i = 0) & = & \theta_\ell^2  \nonumber
\end{eqnarray}
If the sample is contaminated, it is assumed that the sample contains the DNA from two unrelated individuals and that both doses of DNA appear in the observed phenotype of the individual.  Therefore, the following formulas give the probabilities for any given genotype:
\begin{eqnarray} \label{HW_contam}
P(y_{i,\ell} = 0\:|\:z_i = 1) & = & (1-\theta_\ell)^4 \nonumber\\
P(y_{i,\ell} = 1\:|\:z_i = 0) & = & 1 - \theta_\ell^4 - (1-\theta_\ell)^4 \\
P(y_{i,\ell} = 2\:|\:z_i = 1) & = & \theta_\ell^4  \nonumber
\end{eqnarray}
The unobserved variable $\rho$ denotes the contamination rate in the collection. Given $\rho$, each sample in 
the collection is contaminated, independently, with probability $\rho$.  We assume that the DNA 
that contaminates a sample is drawn randomly from the population, independently between samples.  

Figure~\ref{fig:dag1} shows the directed acyclic graph (DAG) for the SSP model. As seen in the DAG, the goal of the model is to use the observed genotype data from each individual at all the loci to infer the allele frequencies in the population, $\theta = (\theta_1,\ldots,\theta_L)$, the contamination rate, $\rho$, and the contamination status of each individual, $z = (z_1,\ldots,z_N)$.  To perform Bayesian inference, prior distributions are placed on $\rho$ and $\theta$.  The 
prior for $\rho$ is a beta distribution with parameters $\alpha$ and $\beta$, and the prior for each $\theta_\ell$ is independently a beta distribution with both parameters equal to $\lambda$. Inference was performed using an MCMC algorithm relying on Gibbs-sample updates for all the unobserved variables.

%%%% The DAG figure
\begin{figure}[h]
\begin{center}
\includegraphics[width=2.5in]{./images/DAG_1.pdf}
\end{center}
\caption[Directed acyclic graph (DAG) representing the SSP model.]{Directed acyclic graph (DAG) representing the SSP model. Black nodes are observed, gray ones are parameters of prior distributions, and unfilled nodes are unobserved variables that are the goal of inference.}
\label{fig:dag1}
\end{figure}
%Derivation of full conditionals
Derivation of the full conditional distributions of $\rho$, $\theta_{\ell}$, and $z_i$ were required to update these unobserved paramenters using Gibbs sampling. The formula for the full conditional of $z_i$ is given Equation~\ref{z_fc}.

\begin{equation} \label{z_fc}
P(z_i|\theta_{\ell},y_i,\rho) = P(z_i|\rho)P(y_i|z_i,\theta)
\end{equation}
$P(z_i|\rho)$ is the likelihood of the $z_i$ value given $\rho$, and $P(y_i|z_i,\theta)$ is the likelihood of the genotype given $z_i$ and $\theta$.  A Bernoulli distribution gives the likelihood of $z_i$, and the likelihood of the genotype is calculated using Equation~\ref{HW_clean} or Equation~\ref{HW_contam}, depending on the value of $z_i$:

\begin{equation} \label{z_lh}
P(z_i|\rho) = \rho^{z_i}(1-\rho)^{1-z_i} 
\end{equation}

\begin{equation} \label{y_lh}
P(y_i|z_i=0,\theta) = \prod_{\ell=1}^{L} P(y_{\ell}|\theta_{\ell},noncontam)
\end{equation}
\begin{equation} \label{y_lhc}
P(y_i|z_i=1,\theta) = \prod_{\ell=1}^{L} P(y_{\ell}|\theta_{\ell},contam)
\end{equation}
After substituting the above equations into Equation~\ref{z_fc}, the full conditional distribution of $z_i$ used in the MCMC modelis given by Equation~\ref{z_simpfc}. Dividing by $R$ normalizes the distribution.
%Full conditional of z
\begin{eqnarray} \label{z_simpfc}
P(z_i=0|\theta_{\ell},y_i,\rho) &=& \frac{(1-\rho)^{1-z_i}\prod_{\ell=1}^{L} P(y_{\ell}|\theta_{\ell},noncontam)}{R} \nonumber \\
P(z_i=1|\theta_{\ell},y_i,\rho) &=& \frac{\rho^{z_i}\prod_{\ell=1}^{L} P(y_{\ell}|\theta_{\ell},contam)}{R} \nonumber \\
R &=& (1-\rho)^{1-z_i}\prod_{\ell=1}^{L} P(y_{\ell}|\theta_{\ell},noncontam) + \rho^{z_i}\prod_{\ell=1}^{L} P(y_{\ell}|\theta_{\ell},contam)
\end{eqnarray}
%Derive full conditional of rho
The formula for the full conditional distribution for $\rho$ is given by:
  \begin{equation} \label{rho_fc}
    P(\rho|z_i) \propto [\prod_{i=1}^{N} P(z_i|\rho)] P(\rho|\alpha,\beta)
  \end{equation}
As shown below, substituting Equation~\ref{z_lh} and the prior beta distribution for $\rho$ can reduce the full conditional distribution of $\rho$ to a beta distribution with parameters $(\alpha + \sum_{i=1}^{N} z_i)$ and $(\beta + N - \sum_{i=1}^{N} z_i)$.

\begin{eqnarray} \label{rho_simpfc}
  P(\rho|z_i) &\propto& [\prod_{i=1}^{N} P(z_i|\rho)] P(\rho|\alpha,\beta) \nonumber \\
  &=& [\prod_{i=1}^{N} \rho^{z_i}(1 - \rho)^{1 - z_i}] \rho^{\alpha - 1}(1 - \rho)^{\beta-1} \nonumber \\
  &=& \rho^{\sum_{i=1}^{N} z_i}(1-\rho)^{N - \sum_{i=1}^{N} z_i} \rho^{\alpha -1} (1-\rho)^{\beta-1} \nonumber \\
  &=& \rho^{\alpha - 1 + \sum_{i=1}^{N} z_i}(1 - \rho)^{\beta -1 + N - \sum_{i=1}^{N} z_i}
\end{eqnarray}
The full conditional distribution of $\theta_{\ell}$ given by:

\begin{equation} \label{theta_fc}
P(\theta_{\ell}|y_{i\ell},z_i) \propto [\prod_{i=\ell}^{N} P(y_{i\ell}|\theta_{\ell},z_i)]P(\theta_{\ell}|\lambda_{\ell})
\end{equation}
Only non-contaminated genotypes are used to calculate allele frequencies.  Therefore, only non-contaminated samples are relevant for the full condition distribution of $\theta_{\ell}$.  The likelihood of $y_{i\ell}$ given $\theta_{\ell}$ and non-contamination, $P(y_{i\ell}|\theta_{\ell},noncontam)$, is calculated assuming that the population is in Hardy-Weinberge Equilibrium. Evaluating $P(y_{i\ell}|\theta_{\ell},noncontam)$ and substituting the prior beta distribution of $\theta_{\ell}$ into Equation~\ref{theta_fc} symplifies the full conditional distribution of $\theta_{\ell}$ to a beta distribution with parameters $(2x_2 + x_1 + \lambda_{\ell})$ and $(2x_0 + x_1 + \lambda_{\ell})$ where $x_j = \sum_{i:z=0}^{N} \delta(y_{i\ell} = j)$. 

\begin{eqnarray} \label{theta_simpfc}
P(\theta_{\ell}|y_{i\ell},z_i) &\propto& [\prod_{i=\ell}^{N} P(y_{i\ell}|\theta_{\ell},z_i)]P(\theta_{\ell}|\lambda_{\ell}) \nonumber \\
&=& [\prod_{i:z=0}^{N} P(y_{i\ell}|\theta_{\ell},noncontam)] \nonumber \\
& & \times \theta_{\ell}^{\lambda_{\ell} - 1}(1-\theta_{\ell})^{\lambda_{\ell} - 1} \nonumber \\
&=& [(1-\theta_{\ell})^2]^{x_0}[2\theta_{\ell}(1-\theta_{\ell})]^{x_1}[\theta_{\ell}^2]^{x_2} \nonumber \\
& & \times \theta_{\ell}^{\lambda_{\ell} - 1}(1 - \theta_{\ell})^{\lambda_{\ell} - 1} \nonumber \\
& \propto& \theta_{\ell}^{2x_0 + x_1 + \lambda_{\ell} - 1}(1-\theta_{\ell})^{2x_0 + x_1 + \lambda_{\ell} - 1}
\end{eqnarray}


\section{Multiple Source Populations Model}
This section details the formulation of the Multiple Source Population (MSP) model, which can be used for detecting contaminated samples from a collection in which the individuals, as well as contaminating DNA, originate from a mixture of population.  As with the SSP model, the collection includes $N$ genetic samples, each one genotyped at $L$ SNPs, but these samples originate from $P$ populations.  Again, samples are assumed to not be in linkage disequilibrium (LD) in each population.  The same variables as in the SSP model represent the alleles, allele frequencies, and genotypes, and $y_{i,\ell}$ is assumed to have the same probability distribution.  Similarly, $z_i$ is used as a contamination indiciator value, and $\rho$ denotes the contamination rate.

In the MSP model, associated with each sample $i$ is another unobserved indicator variable $u_i$ ($u_i\in\{1,\ldots,P\}$), which specifies the population of orgin for sample $i$.  If a sample is not contaminated, the variable $u_i$ takes a single value, and if a sample is contaminated, $u_i$ takes two values, representing the population of origin of each individual in the sample.  The unobserved variable $\pi = (\pi_1,\ldots,\pi_P)$ denotes the mixing proportion from the source of the samples. Given $\pi_k$, a sample in the collection originates from population $k$ with a probability of $\pi_k$.  In the MSP model, we assume that the DNA that contaminates a sample is drawn randomly from the source, independtly between samples, meaning that individuals originating from populations with a higher $\pi$ value are assumed to have a higher probability of being the source of contamination. The model also uses previously collected baseline data, $B$.  The baseline includes the genotype data of a number of individuals whose population of origin is known and is used to estimate the allele frequencies, $\theta = (\theta_1, \ldots, \theta_L)$, in each of the $P$ populatons.  

Figure~\ref{fig:dag2} shows the directed acyclic graph for the MSP model.  The goal of this model is to use the observed baseline data and the observed genotype data from each individual at all the loci to infer the mixing proportions, $\pi = (\pi_1, \ldots, \pi_P)$, population origin of the samples, $u_i = (u_1, \ldots, u_N)$, the contamination rate, $\rho$, and the contamination status of each individual, $z = (z_1, \ldots, z_N)$.  The same prior distributions were placed upon $\rho$ and $\theta$ as in the SSP model, and Dirichlet distributions were placed on parameters $\xi$ on $\pi$. Again, inference was performed using an MCMC algorithm relying on simple Gibbs-sample updates for all the unobserved variables.

%%%% The DAG_2 figure
\begin{figure}[h]
\begin{center}
\includegraphics[width=2.75in]{./images/DAG_2.pdf}
\end{center}
\caption[Directed acyclic graph (DAG) representing the MSP model.]{Directed acyclic graph (DAG) representing the MSP model. Black nodes are observed, gray ones are parameters of prior distributions, and unfilled nodes are unobserved variables that are the goal of inference.}
\label{fig:dag2}
\end{figure}

%Derivation of full conditionals
To preform Gibbs sampling, the full conditional distributions of $\rho$, $\pi$, $u_i$, and $z_i$ were derived. The full conditional distribution of $\rho$ is given by Equation~\ref{rho_simpfc} and is the same distribution used in the SSP model.  The formula for the full conditional of $z_i$ is given Equation~\ref{z_fc2}.

\begin{equation} \label{z_fc2}
P(z_i|\theta_{\ell},y_i,\rho, B) = P(z_i|\rho)P(y_i|z_i,\theta,B)
\end{equation}
The likelihood $P(z_i|\rho)$ is the same as in the SSP model and is given by Equation~\ref{z_lh}. The full conditional distribution of $z_i$ is shown in Equation~\ref{z_simpfc2}. 

\begin{eqnarray} \label{z_simpfc2}
P(z_i=0|\theta_{\ell},y_i,\rho,B) &=& \frac{(1-\rho)^{1-z_i}\sum_{k=1}^{P} \prod_{\ell=1}^{L} P(y_{\ell}|\theta_{\ell,k},noncontam,B)}{R} \nonumber \\
P(z_i=1|\theta_{\ell},y_i,\rho,B) &=& \frac{\rho^{z_i}\sum_{k=1}^{P} \sum_{j=1}^{P} \prod_{\ell=1}^{L} P(y_{\ell,}|\theta_{\ell,k},\theta_{\ell,j},contam,B)}{R} \nonumber \\
R &=& P(z_i=0|\theta_{\ell},y_i,\rho,B) + P(z_i=1|\theta_{\ell},y_i,\rho,B)
\end{eqnarray}

Calculating the likelihood of $y_i$ is more involved than in the SSP model.  A Polya's Urn model was used to model $P(y_i|z_i,\theta,B)$ with a beta binomial distribution.  The distribution when $z_i = 0$ is given by Equation~\ref{clean_bb}.  The variable $n_{\ell,k}$ is the number of samples at the $\ell^{th}$ locus in the $k^{th}$ population and $x_j = \sum_{i}^{N} \delta(y_{i,\ell} = j)$.

\begin{eqnarray} \label{clean_bb}
P(y_{\ell} = 0|clean, \theta_{\ell},B) &=& \frac{(x_0 + 1 + \lambda)(x_0 + \lambda)}{(2n_{\ell,k} + 1)(2n_{\ell,k} + 2)} \nonumber \\
P(y_{\ell} = 2|clean, \theta_{\ell},B) &=& \frac{(x_1 + 1 + \lambda)(x_1 + \lambda)}{(2n_{\ell,k} + 1)(2n_{\ell,k} + 2)} \nonumber \\
P(y_{\ell} = 1|clean, \theta_{\ell},B) &=& 2 \times \frac{(x_0 + \lambda)(x_1 + \lambda)}{(2n_{\ell,k} + 1)(2n_{\ell,k} + 2)}
\end{eqnarray}

Equation~\ref{contam_bb} displays the probability distribution when $z_i = 1$.  There are two separate distributions for if the contamination occurs within the same population or between two different populations. 
\begin{eqnarray} \label{contam_bb}
P(y_{\ell} = 0|contam, \theta_{\ell},B, k=j) &=& \frac{(x_0 + 1 + \lambda)(x_0 + \lambda)(x_0 + 2 + \lambda)(x_0 + 3 + \lambda)}{(2n_{\ell,k} + 1)(2n_{\ell,k} + 2)(2n_{\ell,k} + 3)(2n_{\ell,k} + 4)} \nonumber \\
P(y_{\ell} = 0|contam, \theta_{\ell},B, k \neq j) &=& \frac{(x_0 + 1 + \lambda)(x_0 + \lambda)}{(2n_{\ell,k} + 1)(2n_{\ell,k} + 2)} \times \frac{(x_0 + 1 + \lambda)(x_0 + \lambda)}{(2n_{\ell,j} + 1)(2n_{\ell,j} + 2)} \nonumber \\
P(y_{\ell} = 2|contam, \theta_{\ell},B,k=j) &=& \frac{(x_1 + 1 + \lambda)(x_1 + \lambda)(x_1 + 2 + \lambda)(x_1 + 3 + \lambda)}{(2n_{\ell,k} + 1)(2n_{\ell,k} + 2)(2n_{\ell,k} + 3)(2n_{\ell,k} + 4)} \nonumber \\
P(y_{\ell} = 2|contam, \theta_{\ell},B, k \neq j) &=& \frac{(x_1 + 1 + \lambda)(x_1 + \lambda)}{(2n_{\ell,k} + 1)(2n_{\ell,k} + 2)} \times \frac{(x_1 + 1 + \lambda)(x_1 + \lambda)}{(2n_{\ell,j} + 1)(2n_{\ell,j} + 2)} \nonumber \\
P(y_{\ell} = 1|contam, \theta_{\ell},B) &=& 1 - P(y_{\ell} = 0|contam, \theta_{\ell}, B) - P(y_{\ell} = 2|contam, \theta_{\ell},B)
\end{eqnarray}

The full conditional probablity of $\pi$ was given by Equation~\ref{pi_fc}.  Only non-contaminated samples were used to update the value of $\pi$.  

\begin{equation} \label{pi_fc}
P(\pi|\xi,u) = P(\pi|\xi)P(u|\pi)
\end{equation}

$P(\pi|\xi)$ is represented by a Dirichlet distribution with $\xi = (\frac{1}{P},\ldots,\frac{1}{P})$ and $P(u|\pi)$ by a multinomial distribution shown in Equation~\ref{multi_u}.  After substitution, the full conditional distribution of $\pi$ is given by Equation~\ref{pi_simpfc}, which is a Dirichlet distribution with parameters $(w_1 + \frac{1}{P}, \ldots, w_P + \frac{1}{P})$ where $w_j = \sum_{i}^{N} \delta(u_{i} = j)$.

\begin{equation} \label{multi_u}
P(u|\pi) = \frac{P!}{\prod_{i=1}^{P} w_i!} \prod_{i=1}^{P} \pi_i^{w_i}
\end{equation}

\begin{eqnarray} \label{pi_simpfc}
P(\pi|\xi,u) &\propto& \prod_{i=1}^{P} \pi_i^{\xi-1} \times \frac{P!}{\prod_{i=1}^{P} w_i!} \prod_{i=1}^{P} \pi_i^{w_i} \nonumber \\
&\propto& \prod_{i=1}^{P} \pi_i^{\xi-1} \times \prod_{i=1}^{P} \pi_i^{w_i} \nonumber \\
&=& \prod_{i=1}^{P} \pi_i^{\xi + w_i - 1}
\end{eqnarray}

The full conditional for $u_i$ is given by Equaiton~\ref{u_fc}. There are two separate distributions for non-contaminatd and contaminated samples.  If a sample is not contaminated, the likelihood that $u_i$ is equal to a specific population $k$ is given by $P(u_i = k|\pi) = \pi_k$, and the likelihood of $y_i$ is given by Equation~\ref{clean_bb}.  Therefore, the full conditional distribution of $u_i$ for a non-contaminated sample is given by Equation~\ref{u_simpfc}.

\begin{equation} \label{u_fc}
P(u_i|\pi,y_i,\theta,z_i) = P(u_i|\pi)P(y_i|u_i,\theta,z_i,B)
\end{equation}

\begin{equation} \label{u_simpfc}
P(u_i = k|\pi,z_i=0,y_i,\theta,B) = \frac{\pi_k \times \prod_{\ell=1}^{L} p(y_{\ell}|\theta_{k,\ell},B,nocontam)}{\sum_{k=1}^{P} \pi_k \times \prod_{\ell=1}^{L} P(y_{\ell}|\theta_{k,\ell},B,nocontam)}
\end{equation}

If the sample is contaminated, the likelihood that $u_i$ is equal to ($k$, $j$) is given by $P(u_i = k,j|\pi) = \pi_k \pi_j$.  Order matters in this derivations, so $u_i=k,j$ is different than $u_i=j,k$.  The $P(y_i|u_i,\theta,z_i,B)$ in this scenario is given Equation~\ref{contam_bb}.  Therefore the full conditional distribution of $u_i$ for a contaminated sample is:

\begin{equation} \label{u_simpfc2}
P(u_i = k,j|\pi,z_i=1,y_i,\theta,B) = \frac{\pi_k \pi_j \times \prod_{\ell=1}^{L} p(y_{\ell}|\theta_{k,\ell},\theta_{j,\ell},B,nocontam)}{\sum_{j=1}^P \sum_{k=1}^{P} \pi_k \pi_j \times \prod_{\ell=1}^{L} P(y_{\ell}|\theta_{k,\ell}, \theta_{j,\ell}, B,nocontam)}
\end{equation}

\chapter{Test Simulations Methods}
Simulations were developed and run in R (v.3.1.0) to test the accuracy of the two MCMC models.  The sections below explain the simulation set up for both the SSP and MSP models.

\section{SSP Model Simulation}
The SSP model was tested using \emph{in silico} data generated under Hardy-Weinberg assumptions.  Each simulation had four parameters: $n$, the number of samples; $\rho$, the contamination proportion; $L$, the number of loci in each sample; and $\boldsymbol{\theta}$, the allele frequence at all $L$ loci. For each simulation, $n=200$, meaning that the collection included 200 samples.  $L$ allele frequencies were randomly drawn with replacement from the Feather River Hatchery Spring run Chinook salmon population to create $\boldsymbol{\theta}$ for each simulation \citep{ClementoData, Clementoetal2014}.  Allele frequencies were drawn from real salmon data so that the collection of loci would be similar to those found in nature.  After selecting allele frequencies, \emph{in silico} genotypes were created by sampling the genotype of each locus of each individual using the distribution in Equation~\ref{HW_clean} for non-contaminated individuals and Equation~\ref{HW_contam} for contaminated individuals. $N-N \cdot \rho$ individuals were simulated as "non-contaminated," and $N \cdot \rho$ indivudals individuals were simulated as "contaminated."  The contamination status of each samples was stored in the variable $\boldsymbol{z}$ and was used after the simulations to assess the functioning of the model. Data was simulated with values of $L = (20,60,100,200)$ and $\rho = (0,0.025,0.075,0.2,0.5)$, creating a total of 20 different sets of parameters. One hundred simulations were run for each set of parameters.

The SSP model was run on the 2000 sets of simulated data to test its effectiveness at identifying contaminated samples, estimating the contamination rate, and estimating the allele frequencies in the population.  During the experiments, the hyperparameters for the Beta distribution on $\rho$ ($\alpha$, $\beta$) and on $\theta_{\ell}$ ($\lambda$) were set to 0.5.


\section{MSP Model Simulation}
The MSP model was tested using actual Chinook salmon genotypes with 95 genotyped loci from previously collected baseline data \citep{ClementoData, Clementoetal2014}. Each simulation had three addition parameters to those mentioned in the above section: $\pi$, the mixing proportions for each population; $u$, the population identification of each the individual; and $B$, the baseline data.  Similar to the SSP simulations, $n=200$ for each simulation.  Twenty different sets of mixing proportions observed at different West Coast fisheries at different times of the year were used for $\pi$, and genotypes from the original baseline were sampled using these 20 different sets of $\pi$.  Non-contaminated individuals' genotypes were the sampled genotypes from the original baseline.  We created a number of "contaminated" individuals by combining the genotype of two randomly selected individuals from the original baseline. For each simulation, a new baseline, $B$, was created by removing the randomly sampled genotypes from the original baseline. The data was simulated using the same five values for $\rho$ as in the SSP simulation, creating a total of 100 different sets of parameters.  We ran 20 simulations for each set of parameters.

The MSP model was run on the 2000 sets of simulated data to test its effectiveness at identifying contaminated samples and estimating the contamination rate. As with the SSP model, the hyperparameters $\alpha$, $\beta$, and $\theta_{\ell}$ ($\lambda$) were set to 0.5 during the experiments.

\chapter{Simulation Results}
\section{SSP Simulation Results}

The results of the test simulations proved that the SSP model functions very well on simulated data.  Figure~\ref{rho_allele_SSP}({\em a}) illustrates the models effectiveness at estimating the contamination proportion, $\rho$.  The figure shows a series of boxplots depicting the distribution of the posterior mean of $\rho$ for each of the 20 parameter sets.  In each of the five plots, the box plots are clustered along a line, representing the true values of the contamination proportion.  For all the numbers of loci (20, 60, 100, and 200), the mean of the box plots are very close to the true $\rho$ value and there are very few outliers and very little spread in the distribution, illustrating the model accurately and precisely estimates the contamination proportion of a collection of samples.

The SSP model is also able to successfully estimate the allele frequencies of the population from which the \emph{in silico} collection originated.  The models effectiveness is visually demonstrated by Figure~\ref{rho_allele_SSP}({\em b}), which plots the posterior mean of the $\theta_{\ell}$ against the true allele frequency for a subset of simulations ($L$=20 and $\rho=0.2$).  The data points of Figure~\ref{rho_allele_SSP}({\em b}) are clustered about a dashed line that represents a 1 to 1 ratio, indicating that the estimated frequencies are close to true values.  The results for other values of $L$ and $\rho$ were simular to those displayed in the figure. Table~\ref{allele_CI}, which displays the proportion of true $\theta_{\ell}$ values lying within the 90 percent posterior credible interval (CI), also proves the accuracy of the models allele frequency estimates, as 90 percent of the true allele frequencies consistenly fall within the CI.

%%% The allele frequency CI coverage table
\begin{table}[h]
\caption[Coverage of posterior credible intervals for allele frequencies in SSP model]{Coverage of posterior credible intervals for allele frequencies.  The values in the table represent the proportion of the true allele frequencies that fell within the 90\% posterior credible interval.  The results were calculated using allele frequencies from all 2000 simulations.}
\hrule \hrule
\begin{center}
\input{tables/afreq_ci_coverage}
\label{allele_CI}
\end{center}
\hrule
\end{table}

%%%  RHO AND ALLELE FIGURE
\begin{figure}
\begin{center}
\includegraphics[width=\columnwidth]{./images/rho_and_allele.pdf}
\end{center}
\caption[Posterior mean estimates of $\rho$ and $\theta_{\ell}$ for SSP model]{({\em a}) Boxplots showing the distribution of posterior mean estimates of $\rho$ under varying true values of the proportion of contaminated samples  (listed along the top edge of edge panel) and number of loci (20, 60, 100, 200, as listed along the $x$-axis). Each boxplot summarizes 100 estimates. ({\em b}) Posterior mean estimates (on the $y$-axis) of allele frequencies, $\theta_\ell$,  plotted against the true values (on the $x$-axis).  Results shown are from all 100 simulations with $L=20$ and $\rho=0.2$, so that the individual points can be seen.}
\label{rho_allele_SSP}
\end{figure}

\FloatBarrier
The ultimate goal of the SSP model was to identify the contaminated samples within the collection. Overall, the model accurately identified the majority of the contaminated samples, using a posterior probability $>0.5$ (\ie $P(z_i=1|\mathbf{y})>0.5$) as the criteria for contamination identification. For simulations run with 100 or greater loci, the model correctly identified nearly 100 percent of the contaminated samples, (see Table~\ref{ztable_SSP}({\em a})), while falsely identifying only three non-contamined samples as contaminated over 1000 simulations (see Table~\ref{ztable_SSP}({\em a})).  Even with 60 loci, the model still identified above 98 percent of the contaminated samples, while only falsely labeling less than 0.4 percent of non-contaminted samples as contaminated.  With 20 loci, however, the model is less effective and lacks consistency. True positive rates ranged from 0.578 to 0.942 percent, depending on the value of $\rho$, and false negative rates ranged from 0.001 to 0.061. \\
%%% The false and true positive z_table
\begin{table}[h]
\caption[True positive and false positive rates of contamination identification for SSP model]{True positive and false positive rates for contamination identification across all simulated data sets with given values of $\rho$ and $L$.  {\em (a)}~The fraction of truly contaminated
samples inferred to be contaminated with a posterior probability $>0.5$ of being contaminated, [\ie $P(z_i=1|\mathbf{y})>0.5$].
{\em (b)}~ The fraction of non-contaminated samples inferred to be contaminated with $P(z_i=1|\mathbf{y})>0.5$.}
\hrule\hrule
\mbox{}\\
\input{tables/z_table_p_one_half}
\hrule
\label{ztable_SSP}
\end{table}

\FloatBarrier
\section{MSP Simulation Results}
The results of the test simulations proved that the MSP is also able to identify contaminated samples and estimate the contamination proprotion, $\rho$.  Like Figure~\ref{rho_allele_SSP}, the boxplots in Figure~\ref{MSP_boxplots} depict the distribution of the posterior mean of $\rho$, but in this case for each of five contamination proportion values and 20 different mixing proportions sets.  As with the box plots of the SSP model, those of the MSP model are clustered along a line, representing the true values of the contamination proportion.  For all the mixing proportion sets and contamination proportions, the posterior means are very close to the true $\rho$ value.  In addtion, there are very few outliers and very little spread in the distribution, illustrating the model accurately and precisely estimates the contamination proportion of a collection of samples.


\begin{figure}[h]
\begin{center}
\includegraphics[width=\columnwidth]{./images/mixed_rho}
\end{center}
\caption[Posterior Mean Estimates of contamination proportion MSP Model]{Boxplots showing the distribution of posterior mean estimates of $\rho$ under varying true values of the fraction of contaminated samples  (listed along the top edge of edge panel) and the different fishery proportions (the identity of which are not revelant and not listed on the $x$-axis). Each boxplot summarizes 20 estimates.}
\label{MSP_boxplots}
\end{figure}

Overall, the MSP model accurately identified the majority of the contaminated samples, using a posterior probability $>0.5$ (\ie $P(z_i=1|\mathbf{y})>0.5$) as the criteria for contamination identification.  For all simulations run with each of the 20 mixing proportions, the model correctly at least 98 percent of the contaminated samples, while falsely identifying less than 0.3 percent of clean samples as contaminated (see Table~\ref{ztable_MSP}).  No non-contaminated samples were falsely identified as contaminated when the contamination proportion was 0 or 0.025; however, in most simulation senarios with larger $\rho$ values (\ie $\rho = (0.075, 0.2, 0.5)$), there where a few false positives.  There seems to be little difference in the true positive rates between the scenarios run with different contamination proportions.

\begin{table}[h]
\caption[True positive and false positive rates of contamination identification of MSP Model]{True positive and false positive rates.  {\em (a)}~Across all simulated 
data sets with given values of $\rho$ and the fishery proportions used, the fraction of truly contaminated
samples inferred to have posterior probability $>0.5$ of being contaminated, [\ie $P(z_i=1|\mathbf{y})>0.5$].
{\em (b)}~Fraction of non-contaminated samples with $P(z_i=1|\mathbf{y})>0.5$.}
\hrule\hrule
\mbox{}\\
\input{tables/mixed_z_table}
\hrule
\label{ztable_MSP}
\end{table}

\chapter{Conclusion}
This paper presented a model for identifying contaminated samples genotyped using single nucleotide polymorphisms (SNPs).  Unlike highly polymorphic microsatellites, SNPs only have two alleles; therefore, contaminated samples cannot be dected by simply identifying samples that are genotyped with more than two alleles at a specific locus.  Models have been developed to identify contaminated samples of human DNA genotyped with SNPs, but few studies have investigated ways to detect contaminated samples of DNA from non-model organisms.  The Baysian models described in this paper identify contaminated SNP samples given an unknown rate of contamination.  Two different models were developed in this study: (1) the SSP model, tailored for identifying contamination in collections of samples used to estimate allele frequencies of a single population and (2) the MSP model, tailored for detecting contamination in collections of samples used for the population identification of organisms sampled from a mixture of multiple populatins.  Both models rely on the assumption that populations sampled are not in linkage disequilibrium, that each sample's contamination status is idendepent of those of other samples, and that the DNA that contaminates a sample is drawn randomly from the population.

The results of the simulation tests showed that both the SSP and MSP models are able to accurately estimate the contaminateion rate, $\rho$, and correctly identify the contamination status of the vast majority of the samples.  In addition the SSP model is able to effectively estimate the allele frequencies of the population from which the collection of samples originated.  Notably, the SSP model's effectiveness decreased with a decreasing number of loci tested.  The model was most effective with 60 and above loci, but was not able to accurately identify the contamination status with samples genotyped with 20 loci, most likely because the information contained in a small number of loci is not sufficient to estimate the contamination status.  The MSP model was tested with real salmon SNP data and the number of loci was not varied; however, it is reasonable to assume that the MSP model requires a similar threshold for number of loci as both models are based off similar formulas and assumptions.  The MSP model was accurate for a variaty of mixing proportions found in nature, indicating that the mixture composition does not effect the model's performance.

The Bayesian models described in this paper assumes that a contaminated sample includes two individuals and that the genotypes of both individuals will be registered at every loci.  These assummptions make these models ideal for identifying contamination that occurs as a result of laboratory error that has accidentally pooled two DNA samples.  Other sources of contamination, occuring both the collection and laboratory use of DNA, may violoate these assumptions.  For example, small amounts of aerosolized DNA could contaminate other DNA samples; however, this small amount of DNA may not be registered in all the genotyped loci.  The partial contamination of samples would most likely reduce the accuracy of the models.  The SSP and MSP models are a good base test for identifying certain types of contamination, but further exploration should be done to develop models that consider varying percentages of affected loci.